### ISSUE TITLE
rclone mount with WebDAV file write error

### ISSUE THREAD

**Issue by Niddrym** [rclone/rclone#2603]
#### What is the problem you are having with rclone?
While using rclone mount remote: when I attempt to write files I'm receiving the following error:
2018/09/23 12:45:59 ERROR : firstScript: Failed to copy: Update mkParentDir failed: 403                                                                
2018/09/23 12:45:59 ERROR : firstScript: WriteFileHandle.New Rcat failed: Update mkParentDir failed: 403                                               
2018/09/23 12:45:59 ERROR : firstScript: WriteFileHandle.Flush error: Update mkParentDir failed: 403 

#### What is your rclone version (output from `rclone version`)
rclone v1.43.1
- os/arch: linux/amd64
- go version: go1.11

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)
MX Linux 17 (debian based) 64 bit

####  Which cloud storage system are you using? (eg Google Drive)
WedDAV provided by Blackboard

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)
rclone mount remote:
then cp 
or vim to edit files
or save using any basic text editor

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)
[bb.log](https://github.com/ncw/rclone/files/2408927/bb.log)

**Comment by Cnly** [comment]
Hi! I took a quick look at your log and found that you can read files on remote without any problem. The `403` error you've got was probably from the remote and means `Forbidden`. Can you confirm that you have write access to the remote? Thanks!

**Comment by Niddrym** [comment]
I do have write access, I'm actually able to mount the drive using davfs2 and read and write successfully.  I've also mounted the drive using windows explorer network drive functionality though I doubt that information is particularly helpful.

**Comment by Niddrym** [comment]
Just did a little digging out curiosity, I'm able to successfully copy if the parent directory doesn't exist.  For example if I create a new sub-directory test/firstScript.
I'm wondering if the issue maybe in the mkdir function in webdav.go where, i suspect maybe in this implementation of this webdav server returns of 403 instead of a logical 405 or 406 upon failure to create a parent directory:
```
		if apiErr.StatusCode == http.StatusMethodNotAllowed || apiErr.StatusCode == http.StatusNotAcceptable {
			return nil
		}
```

**Comment by Cnly** [comment]
Yes, that info is very helpful. I just looked at the code and did find something suspicious.

rclone attempts to create the parent directories (except root) for a file even if they already exist. ([code](https://github.com/ncw/rclone/blob/master/backend/webdav/webdav.go#L608)) And your `403` probably comes from this behaviour. By the [standard](https://tools.ietf.org/html/rfc4918#section-9.3.1), the remote should return `405` for existing locations, but maybe not so when the user's permission is not enough:

> 403 (Forbidden) - This indicates at least one of two conditions: 1)
> the server does not allow the creation of collections at the given
> location in its URL namespace, or 2) the parent collection of the
> Request-URI exists but cannot accept members.

> 405 (Method Not Allowed) - MKCOL can only be executed on an unmapped
> URL.

I also tested with my Apache2 WebDAV and it returns `405` when the directory exists but the user has no permission to make new directories.

Maybe we need to change rclone's webdav backend to make it more compatible? @ncw

**Comment by Cnly** [comment]
> Just did a little digging out curiosity, I'm able to successfully copy if the parent directory doesn't exist. For example if I create a new sub-directory test/firstScript.
> I'm wondering if the issue maybe in the mkdir function in webdav.go where, i suspect maybe in this implementation of this webdav server returns of 403 instead of a logical 405 or 406 upon failure to create a parent directory:
> 
> ```
> 		if apiErr.StatusCode == http.StatusMethodNotAllowed || apiErr.StatusCode == http.StatusNotAcceptable {
> 			return nil
> 		}
> ```

You replied a few minutes before I did. I'm now wondering how many implementations are like this.

**Comment by Niddrym** [comment]
I just tested adding the 403 or condition to that if statement and I'm able to write new files, just fine.  So that will correct the initial bug of the ticket.  

However I'm getting a new error 412 with overwriting files.  I'll keep debugging, but maybe you guys could have a look?

[bb2.log](https://github.com/ncw/rclone/files/2409594/bb2.log)

**Comment by Niddrym** [comment]
To correct my issue with overwriting existing files, in the Move function of WebDAV, I need to use the method PUT instead of MOVE to overwrite and this corrects the errors, and maintains links in the WebDAV implementation.  However a server side only move still requires the MOVE method to be used.  Unfortunately I'm not too familiar with go, would it be possible to add an additional vendor to support using PUT to overwrite?   I'm putting my modified move function in this comment.

```
func (f *Fs) Move(src fs.Object, remote string) (fs.Object, error) {
	info, err := f.readMetaDataForPath(remote, "0")
	if info != nil && err == nil {
		return f.copyOrMove(src, remote, "PUT")
	}
	return f.copyOrMove(src, remote, "MOVE")
}
```

**Comment by burbanczyk** [comment]
Hello, @Cnly any progress on this bug? 
I'm getting the exact same error. 
```
2020/07/05 14:31:22 ERROR : styles/_variables.less: Failed to copy: Update mkParentDir failed: 403 Forbidden
2020/07/05 14:31:22 ERROR : styles/_variables.less: WriteFileHandle.New Rcat failed: Update mkParentDir failed: 403 Forbidden
2020/07/05 14:31:22 ERROR : styles/_variables.less: WriteFileHandle.Flush error: Update mkParentDir failed: 403 Forbidden
2020/07/05 14:31:22 ERROR : IO error: Update mkParentDir failed: 403 Forbidden
```
@Niddrym did you find a workaround?

**Comment by bderiso** [comment]
I'm getting this error in version 1.52.2 (the latest available installed via home-brew on macOS) trying to **sync** to sharepoint.

**Comment by ncw** [comment]
Give this a go, it fixes a very similar issue, it might fix yours too

[v1.52.2-223-g0f165122-fix-4428-webdav-beta](https://beta.rclone.org/branch/v1.52.2-223-g0f165122-fix-4428-webdav-beta/) on branch [fix-4428-webdav](https://github.com/rclone/rclone/tree/fix-4428-webdav) (uploaded in 15-30 mins)

**Comment by burbanczyk** [comment]
@ncw 
Thanks, this solved the issue!

**Comment by ncw** [comment]
Great!

That is merged to [the latest beta](https://beta.rclone.org) now and will be released in 1.53
